rockchip: Add support for RK3566 FriendlyElec NanoPi R3S

kz26 opened the pull request:
This commit adds support for the FriendlyElec NanoPi R3S.

CPU: Rockchip RK3566, Quad-core Cortex-A55
RAM: 2GB LPDDR4X
Ethernet: GMAC RTL8211F GbE, PCIe R8111H GbE
USB3.0 Host: Type-A x1
Storage: MicroSD Slot x 1, and optional on-board 32GB eMMC
Debug Serial Port: 3.3V TTL, 3-pin 2.54mm pitch connector, 1500000 bauds
LED: LED x 3
RTC: One low-power RTC, supports backup battery input

Both GbE controllers are working (WAN eth0, LAN eth1).
Appropriate LAN/WAN interface assignments and MAC address generation.
All three LEDs are working.
USB appears to be working and has been tested with mass storage.

Installation - microSD:

-Uncompress the OpenWRT sysupgrade.img.gz
-Write image to microSD card using dd or similar tool

Installation - eMMC:

-Boot from microSD
-Uncompress the OpenWRT sysupgrade.img.gz
-Flash to eMMC : dd if=x.img of=/dev/mmcblk0
-sync
-Remove microSD card
-Reboot

Signed-off-by: Kevin Zhang <kevin@kevinzhang.me>

hauke commented on target/linux/rockchip/patches-6.6/401-v6.6-arm64-dts-rockchip-rk3566-Add-Nanopi-R3S.patch:
The version number in the patch name should be  indicate in which upstream kernel this patch is included. you gave kernel 6.6. here, but that is unlikely because we are using kernel 6.6 as a base. 

Same for the u-boot patch.

kz26 replied:
Renamed, thanks. As this is a new board it is neither in u-boot nor kernel upstream.

PolynomialDivision commented:
Please have a look at https://github.com/openwrt/openwrt/pull/16462.
The PR uses 2 commits. So one extra commit for the uboot.

kz26 commented:
> Please have a look at #16462. The PR uses 2 commits. So one extra commit for the uboot.

Done

hauke commented:
@1715173329 could you have a look please.

1715173329 commented:
https://github.com/1715173329/imoutowrt/blob/master-rockchip-r3s/target/linux/rockchip/files/arch/arm64/boot/dts/rockchip/rk3566-nanopi-r3s.dts

I have a "cleaned up" version there, will submit it to upstream kernel these days.

1715173329 commented on target/linux/rockchip/armv8/base-files/etc/board.d/02_network:
```diff
@@ -66,6 +67,7 @@ rockchip_setup_macs()
 		wan_mac=$(get_mac_binary "/sys/bus/i2c/devices/2-0051/eeprom" 0xfa)
 		lan_mac=$(macaddr_setbit_la "$wan_mac")
 		;;
+	friendlyarm,nanopi-r3s|\
```
please move it above `friendlyarm,nanopi-r4s-enterprise`

kz26 replied:
What would be the reason for this? AFAIK the R3S does not have unique MAC addresses.

1715173329 replied:
Just reorder it alphabetically, like this:

```
	friendlyarm,nanopi-r3s|\
	friendlyarm,nanopi-r5c|\
	friendlyarm,nanopi-r6c|\
	friendlyarm,nanopi-r6s)
		wan_mac=$(macaddr_generate_from_mmc_cid mmcblk*)
		lan_mac=$(macaddr_add "$wan_mac" 1)
		;;
	friendlyarm,nanopi-r4s-enterprise)
		wan_mac=$(get_mac_binary "/sys/bus/i2c/devices/2-0051/eeprom" 0xfa)
		lan_mac=$(macaddr_setbit_la "$wan_mac")
		;;
```

1715173329 commented on package/boot/uboot-rockchip/patches/109-rockchip-add-FriendlyElec-NanoPi-R3S.patch:
```diff
@@ -0,0 +1,758 @@
+--- /dev/null	2024-10-08 20:38:15.835000058 -0400
```
please refresh uboot patch with `make package/uboot-rockchip/{clean,refresh} V=s`
maybe this patch should start with `100-`.

1715173329 commented on target/linux/rockchip/patches-6.6/401-arm64-dts-rockchip-rk3566-Add-Nanopi-R3S.patch:
```diff
@@ -0,0 +1,658 @@
+--- /dev/null
++++ b/arch/arm64/boot/dts/rockchip/rk3566-nanopi-r3s.dts
```
It seems to be a mix of upstream nanopi-r5s dts and vendor nanopi-r3s dts.

However the regulator parts is simpler on r3s, there's no `vcc3v3_pcie` and `vcc5v0_usb_*`. For other parts, like `usb2phy1*` is disabled by default so you don't need to disable them again.

I have sent nanopi r3s support to upstream kernel, could you use that version instead?
https://lore.kernel.org/linux-rockchip/20241020173946.225960-2-cnsztl@gmail.com/

Addtionally, please create a separated patch to add OpenWrt's LED aliases, fyi:
https://github.com/openwrt/openwrt/blob/913368a223c3818d2b5b60b06820532038be3467/target/linux/rockchip/patches-6.6/118-arm64-dts-rockchip-Update-LED-properties-for-Radxa-Ro.patch#L19-L22

kz26 replied:
Done - thanks for the feedback. I have refactored the patches using your upstream-submitted DTS, added a separate patch for the LED aliases, and realphabetized 02_network. I've also tested a new build with the aforementioned changes and can confirm that things work as expected.

1715173329 commented on package/boot/uboot-rockchip/patches/109-rockchip-add-FriendlyElec-NanoPi-R3S.patch:
```diff
@@ -0,0 +1,666 @@
+--- /dev/null
```
please renumber this patch to 100, i.e. `package/boot/uboot-rockchip/patches/100-rockchip-add-FriendlyElec-NanoPi-R3S.patch`

1715173329 commented on package/boot/uboot-rockchip/patches/109-rockchip-add-FriendlyElec-NanoPi-R3S.patch:
```diff
@@ -0,0 +1,666 @@
++CONFIG_SPL_OF_CONTROL=y
++CONFIG_OF_LIVE=y
++CONFIG_OF_SPL_REMOVE_PROPS="clock-names interrupt-parent assigned-clocks assigned-clock-rates assigned-clock-parents"
++CONFIG_SPL_DM_WARN=y
```
```suggestion
```
please drop this config, see upstream: https://lore.kernel.org/u-boot/20240802234846.2654274-1-jonas@kwiboo.se/

kz26 replied:
Thanks, done.

1715173329 commented on package/boot/uboot-rockchip/patches/109-rockchip-add-FriendlyElec-NanoPi-R3S.patch:
```diff
@@ -0,0 +1,666 @@
++CONFIG_SYSRESET=y
++CONFIG_USB=y
++CONFIG_USB_XHCI_HCD=y
++CONFIG_USB_EHCI_HCD=y
++CONFIG_USB_EHCI_GENERIC=y
++CONFIG_USB_OHCI_HCD=y
++CONFIG_USB_OHCI_GENERIC=y
```
```suggestion
```

this board does not have ehci and ohci interfaces, please remove them

kz26 replied:
Thanks, done

1715173329 commented on target/linux/rockchip/patches-6.6/401-1-arm64-dts-rockchip-rk3566-Add-Nanopi-R3S.patch:
```diff
@@ -0,0 +1,567 @@
+--- /dev/null
```
please renumber this to 127, i.e. `target/linux/rockchip/patches-6.6/127-arm64-dts-rockchip-rk3566-Add-Nanopi-R3S.patch`

1715173329 commented on target/linux/rockchip/patches-6.6/401-2-arm64-dts-rockchip-rk3566-Nanopi-R3S-update-LED.patch:
```diff
@@ -0,0 +1,14 @@
+--- a/arch/arm64/boot/dts/rockchip/rk3566-nanopi-r3s.dts
```
please renumber this to 128, i.e. `target/linux/rockchip/patches-6.6/128-arm64-dts-rockchip-rk3566-Nanopi-R3S-update-LED.patch`

kz26 replied:
Thanks, done

1715173329 reviewed (commented):
otherwise lgtm, thanks

Commit:
uboot-rockchip: add support for NanoPi R3S

This commit adds u-boot support for the NanoPi R3S.

Signed-off-by: Kevin Zhang <kevin@kevinzhang.me>
Link: https://github.com/openwrt/openwrt/pull/16738
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Commit:
rockchip: Add support for RK3566 FriendlyElec NanoPi R3S

This commit adds support for the FriendlyElec NanoPi R3S.

CPU: Rockchip RK3566, Quad-core Cortex-A55
RAM: 2GB LPDDR4X
Ethernet: GMAC RTL8211F GbE, PCIe R8111H GbE
USB3.0 Host: Type-A x1
Storage: MicroSD Slot x 1, and optional on-board 32GB eMMC
Debug Serial Port: 3.3V TTL, 3-pin 2.54mm pitch connector, 1500000 bauds
LED: LED x 3
RTC: One low-power RTC, supports backup battery input

Both GbE controllers are working (WAN eth0, LAN eth1).
Appropriate LAN/WAN interface assignments and MAC address generation.
All three LEDs are working.
USB appears to be working and has been tested with mass storage.

Installation - microSD:

-Uncompress the OpenWRT sysupgrade.img.gz
-Write image to microSD card using dd or similar tool

Installation - eMMC:

-Boot from microSD
-Uncompress the OpenWRT sysupgrade.img.gz
-Flash to eMMC : dd if=x.img of=/dev/mmcblk0
-sync
-Remove microSD card
-Reboot

Signed-off-by: Kevin Zhang <kevin@kevinzhang.me>
Link: https://github.com/openwrt/openwrt/pull/16738
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Commit:
rockchip: Add LED aliases for FriendlyElec NanoPi R3S

This commit adds OpenWRT's LED aliases to the board DTS
for showing system status on the NanoPi R3S.

Signed-off-by: Kevin Zhang <kevin@kevinzhang.me>
Link: https://github.com/openwrt/openwrt/pull/16738
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
